### ISSUE TITLE
-principal causes assertion failure in type checker

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7838]
**Original bug ID:** 7838
**Reporter:** @mmottl
**Assigned to:** @garrigue
**Status:** resolved (set by @trefis on 2018-08-29T10:01:58Z)
**Resolution:** fixed
**Priority:** normal
**Severity:** major
**Version:** 4.07.0
**Category:** typing
**Monitored by:** @nojb @gasche @mmottl

## Bug description

Consider the following code in file "foo.ml":

```
module Make (X : sig val f : [ `A ] -> unit end) = struct
  let make f1 f2 arg = match arg with `A -> f1 arg; f2 arg
  let f = make X.f (fun _ -> ())
end
```

Executing "ocaml -principal foo.ml" will yield:

    Fatal error: exception File "typing/btype.ml", line 177, characters 9-15: Assertion failed

The problem will only appear with "-principal" and does not seem to exist in versions before OCaml 4.07.0.

**Comment by vicuna** [comment]
**Comment author:** @trefis

I had a look this afternoon and I believe this is an old bug in Ctype.copy that was uncovered by the introduction of more calls to generalize_structure in type_cases (by https://github.com/ocaml/ocaml/pull/1609/commits/d43ccfc5d48ce86202de3c7c4f58317602d6d417 ).

I'm guessing (but I don't really understand that part of the code) the issue is that in the Tvariant case, we're mutating more.desc, without necessarily having saved it before (e.g. when more.desc is Tnil and partial = Some _).
I'll let someone who knows that part look into the issue further.

**Comment by vicuna** [comment]
**Comment author:** @garrigue

Please see https://github.com/ocaml/ocaml/pull/2001 for a fix.

Thomas' analysis is correct.
The introduction of the partial parameter in Ctype.copy created a wrong case where more was modified without being copied (since it was not generalized).

As to why it was not triggered before, I believe this is more related to the fact that partial copy is only invoked in type_cases, and this was limited to pattern-matches containing explicit GADT constructors, or polymorphic variants both in the pattern and in the argument type, which is not the case here. In 4.07, this is called always with in principal mode.

**Comment by vicuna** [comment]
**Comment author:** @garrigue

Sorry, the analysis of why the bug didn't occur before is wrong: the condition for partial_instance has not changed, just the place where it is computed.
On the other hand I do not see how the extra generalizations in type_cases could be related to that: they only occur after the copy.
Anyway, this original code was incorrect, and I believe this is just by chance that this was not triggered before.

**Comment by vicuna** [comment]
**Comment author:** @trefis

https://github.com/ocaml/ocaml/pull/2001 has been merged on 4.07 and cherry-picked onto trunk.
